src/Microsoft.AspNetCore.SignalR.Client/HubConnection.cs

```diff
@@ -25,11 +26,18 @@ public class HubConnection : IDisposable
         private readonly HubBinder _binder;
 
         private readonly CancellationTokenSource _readerCts = new CancellationTokenSource();
-        private readonly ConcurrentDictionary<string, InvocationRequest> _pendingCalls = new ConcurrentDictionary<string, InvocationRequest>();
+        private readonly CancellationTokenSource _connectionActive = new CancellationTokenSource();
+
+        // We need to ensure pending calls added after a connection failure don't hang. Right now the easiest thing to do is lock.
+        private readonly object _pendingCallsLock = new object();
+        private readonly Dictionary<string, InvocationRequest> _pendingCalls = new Dictionary<string, InvocationRequest>();
+
         private readonly ConcurrentDictionary<string, InvocationHandler> _handlers = new ConcurrentDictionary<string, InvocationHandler>();
 
         private int _nextId = 0;
 
+        public Task Completion { get; }
```

Comment:
Yeah, this is a pattern that is developing in other places, we need to decide how we want to express it. You can easily implement an OnConnectionClosed event handler with ContinueWith.